envoy/api/v2/route/route.proto

```diff
@@ -97,6 +97,12 @@ message VirtualHost {
   // [#not-implemented-hide:]
   // Return a 401/403 when auth checks fail.
   auth.AuthAction auth = 9;
+
+  // [#not-implemented-hide:]
+  // Specifies the number of proxies between the origin client and this Envoy instance that
+  // are known to append an accurate source IP address to the X-Forwarded-For HTTP request
+  // header. The default is 1.
+  uint32 num_trusted_ingress_proxies = 12;
```

Comment:
If this defaults to 1, I think we should use the wrapped type, and put a constraint that it is >= 1?

Reply:
Also, per @wora I might try to better clarify what this is doing, and potentially but "xff" in the name somehow.

Reply:
Based on how the field is defined, we'll need to support a value of zero to handle the most common case. The default of 1 is just for backward compatibility with what appears to be a bug in how Envoy handles XFF (depending on how I interpret the current documentation).

The [XFF docs](https://www.envoyproxy.io/docs/envoy/latest/configuration/http_conn_man/headers.html#config-http-conn-man-headers-x-forwarded-for) docs say that "only the last address (furthest to the right) can be trusted." Read in isolation, that suggests that Envoy defaults to assuming a trusted HTTP proxy in front of it. However, other text in that section mentions that "Envoy uses the final XFF contents to determine whether a request originated externally or internally," where "final" seems to mean after Envoy has appended its immediate client's address. So it's not altogether clear from that section whether the current behavior is for Envoy to:

1. Treat the unmodified ingress XFF as untrusted, but trust the immediate client's address that Envoy appends to XFF (which would match how other HTTP implementors handle incoming XFF)

Or:

2. Treat the last non-RFC1918 in the unmodified ingress XFF (before Envoy has modified the XFF at all) as trusted (which would match my reading of the current docs, but it seems like a bug compared to how other implementations handle XFF).

If (1) is the right interpretation, then I'll make the default for this new field zero (and rewrite the XFF docs to clarify whether they're talking about the pre-modification or post-modification XFF).

Reply:
> Read in isolation, that suggests that Envoy defaults to assuming a trusted HTTP proxy in front of it. 

This depends on the "use_remote_address" setting. Can you take a look at that and see if that makes sense? (Basically if "use_remote_address" is set, Envoy will append to XFF and use that to determine trusted status, if it's not set, it won't append, and use XFF directly).

Reply:
After mapping out the various potential configurations, I think it makes sense to have this new setting default to 0.

Let's call this new setting `xff_num_trusted_hops` (I'll update the PR to match). The semantics for determining the trusted origin client address would be:
1. First, either do or don't append the immediate client's IP address to XFF, depending on the use_remote_address setting.
2. Find the first non-RFC1918 address from the end of the XFF.
3. (New step) Move forward _xff_num_trusted_hops_ hops to the left from that location, and the result is the trusted origin client address.

Reply:
I'm not sure I follow step (2), given that addresses cannot be implicitly trusted just because they are RFC1918. IMO isn't just (1) and (3) sufficient with a default of 0? Basically, assume that the address furthest to the right is the trusted address, unless xff_num_trusted_hops is set and then we go to the left?

Reply:
My motivation in step (2) is to preserve the current semantics from the [XFF docs](https://www.envoyproxy.io/docs/envoy/latest/configuration/http_conn_man/headers.html#config-http-conn-man-headers-x-forwarded-for), "More specifically, the first external (non RFC1918) address from the _right_ is the only trustable addresses."

Reply:
I think that's a bug in the docs effectively. Can you review the code to check me and make sure we are on the same page? This is some of the code that is most scary (and confusing) from a security perspective so want to make sure we are on the same page:
https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_utility.cc#L48

Reply:
On the other hand, the [current implementation](https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_utility.cc#L53-L70) doesn't seem to implement the semantics described in the docs. The implementation does this:
1. If use_remote_address is set, the trusted remote address is the address of Envoy's immediate client, regardless of whether it's RFC1918.
2. If use_remote_address is not set, the trusted remote address is the last address in XFF, again regardless of whether it's RFC1918. If there's no XFF, though, the trusted remote address is the address of Envoy's immediate client.

Which should I preserve, the semantics described in the docs or the current behavior of the current implementation? The implementation is more consistent with how I've seen other HTTP implementors handle XFF, but the documentation is so specific about differentiating public and RFC1918 addresses that I can't tell which one should be considered authoritative.

Reply:
@mattklein123 I just saw your reply after I posted mine. I'll build the new feature on top of the implementation semantics, and change the docs to match.